### DISCUSSION TITLE
how to avoid black edge when useing compose in

### DISCUSSION THREAD

**Discussion by mike-kong** [ImageMagick/ImageMagick#6830] [Help] [answered]
### ImageMagick version

7

### Operating system, version and so on

windows

### Description

convert "backgroud.png" -compose in "cover.png"   -gravity center -composite  png32:"output.png"  2>&1

when using compose in ,the output.png will contains translucent black edges.
The semi transparent edge of the cover generates a black semi transparent edge after being composed in.
backgroud.png：
![background](https://github.com/ImageMagick/ImageMagick/assets/92786267/a2ecc9bb-35c4-460f-aa84-1ad4ea1666c4)
cover.png
![cover](https://github.com/ImageMagick/ImageMagick/assets/92786267/da44ea8d-35c1-4923-988c-6e8f79ee1a95)
output.png
![output](https://github.com/ImageMagick/ImageMagick/assets/92786267/e5b873ae-721e-4139-9355-8aafa269c117)
the difference, i hope to avoid translucent black edges or to delete it
![difference](https://github.com/ImageMagick/ImageMagick/assets/92786267/fae1129d-d18a-4cda-bff0-5fdc3a1faaa2)

**Comment by snibgo** [comment]
The documentation at https://imagemagick.org/Usage/compose/#dstin says:
> All 'Dst_In and 'Src_In' methods actually does is multiply the alpha channels of the two images. The color of the appropriate image (according to the method) is preserved.

That isn't true. It also multiplies colour values by alpha. I don't think it should do so. Using "-channel RGB" doesn't fix the problem.

Anyhow, a workaround, to simply multiply the alpha channels of the two images (provided they are the same size), is:
```
magick background.png cover.png ( -clone 0,1 -alpha extract -compose Multiply -composite ) -delete 0 -alpha off -compose CopyOpacity -composite out.png
```
![out](https://github.com/ImageMagick/ImageMagick/assets/33812368/c7ebd283-c98e-47cf-9476-8225260512cf)

**Reply by mike-kong** [reply, marked as answer]
great  great  great great  great job. thank you very much much  much  much  .
Since English is not my native language, the image magick documentation is simply too difficult for me. Thank you for your reply.
Referring to your code, I have successfully synthesized the functions I need
`magick  background.png cover.png ( -clone 0,1  -compose clear -gravity center -composite )  ( -clone 1,2 +swap -compose over -gravity center -composite ) -delete 1,2 ( -clone 0,1  -compose IN -gravity center -composite -alpha extract ) -delete 0 -alpha off -compose CopyOpacity  -composite png32:"output.png"  2>&1`
